ZukoAzula2016/src/org/team1540/zukoazula/Shooter.java

```diff
@@ -30,44 +32,101 @@
-                states.getIsState("passive").toFloat(indexerIntakeSpeed.negatedIf(states.getIsState("backward")), indexerPassiveSpeed));
-
-        BooleanInput shouldSpinUp = setupRollersForSpinup(spinup, actuallyFiring);
-
-        shooter.setup(shouldSpinUp);
-
-        fire.and(shooter.isUpToSpeed).onPress().send(actuallyFiring);
+        
+        StateMachine flywheelTarget = new StateMachine(0, "off", "low", "high");
+        StateMachine flywheelActual = new StateMachine(0, "off", "low", "high");
+        
+        StateMachine shooterStates = new StateMachine(0, 
+                "passive", // do nothing
+                "ejecting", // when the ball is being ejected from the intake mechanism
+                "intaking", // when the intake mechanism is intaking a ball
+                "loaded", // when a ball is loaded in the intake mechanism
+                "cocking", // when a ball is being cocked so that it is not in contact w/ the flywheel 
+                "spinup",  // when the flywheel is spinning up to maximum speed
+                "firing"); // when the ball is firing
+        
+        TalonExtendedMotor flywheel = makeLinkedTalons();
+        flywheel.modEncoder().configureEncoderCodesPerRev(125*15);
+        flywheel.modGeneralConfig().getBrakeNotCoast().set(false);
+        FloatInput flywheelSpeed = flywheel.modEncoder().getEncoderVelocity().absolute();
+        FloatOutput flywheelSimple = flywheel.simpleControl().addRamping(ZukoAzula.mainTuning.getFloat("Shooter Flywheel Ramping", 0.01f).get(), 
+                FRC.constantPeriodic);
+        
+        BooleanInput inhaleButton = ZukoAzula.controlBinding.addBoolean("Shooter Intake");
+        BooleanInput exhaleButton = ZukoAzula.controlBinding.addBoolean("Shooter Eject");
+        BooleanInput prefireButton = ZukoAzula.controlBinding.addBoolean("Shooter Spinup");
+        BooleanInput fireButton = ZukoAzula.controlBinding.addBoolean("Shooter Fire");
+        BooleanInput cancelShooterButton = ZukoAzula.controlBinding.addBoolean("Shooter Cancel Action");
+        
+        // Behavior
+        
+        flywheelSimple.setWhen(0.0f, FRC.duringTele.and(flywheelTarget.getIsState("off")));
+        flywheelSimple.setWhen(-0.068f, FRC.duringTele.and(flywheelTarget.getIsState("low")));
+        flywheelSimple.setWhen(1.0f, FRC.duringTele.and(flywheelTarget.getIsState("high")));
+        
+        flywheelActual.setStateWhen("off", flywheelSpeed.atMost(ZukoAzula.mainTuning.getFloat("Shooter Flywheel Maximum Off Speed", 10.0f)).onPress());
+        flywheelActual.setStateWhen("low", flywheelSpeed.inRange(ZukoAzula.mainTuning.getFloat("Shooter Flywheel Minimum Low Speed", 10.0f), 
+                ZukoAzula.mainTuning.getFloat("Shooter Flywheel Maximum Low Speed", 1000.0f)).onPress());
+        flywheelActual.setStateWhen("high", flywheelSpeed.atLeast(ZukoAzula.mainTuning.getFloat("Shooter Flywheel Minimum High Speed", 2000.0f)).onPress());
```

Comment:
I'm not 100% certain that a StateMachine is the right choice here. Any particular reason that you didn't use something more functional (as opposed to imperative)?